Add support for adjusting size of skin elements

peppy opened the pull request:
Of note, a few components will still show the drag handles even though adjustments won't work. This is because there's some weird cases like `Update() { Size = ... }`. I don't see this as a deal-breaker or something we need to address immediately.

---

Dragging a skin elements from an edge will now adjust size, not scale. This allows resizing elements which support it without stretching them (which looked really bad).

Support is limited for now, but we can easily roll this out to more components in the future.

| Before | After |
| :---: | :---: |
| ![CleanShot 2023-11-10 at 05 37 19](https://github.com/ppy/osu/assets/191335/fbebe760-e2de-40ae-9cc1-8d1102bfbec2) | ![CleanShot 2023-11-10 at 05 39 58](https://github.com/ppy/osu/assets/191335/67f1d580-53b5-4e80-b38b-e133412b53c8) |

https://github.com/ppy/osu/assets/191335/a15d7697-3b74-4266-a209-c0b95d90f11d

peppy committed:
Add `Size` to serialised components of a `SerialisedDrawableInfo`

peppy committed:
Change selection handling to adjust `Size` instead of `Scale` for edge nodes

peppy committed:
Reset size along with scale for relative items

peppy committed:
Disable resizing of `LegacySongProgress`

Because it looks bad.

peppy committed:
Reset scale per axis

bdach committed:
Fix compose select box test failures

bdach commented:
I've fixed a few trivial test fails but I'm not sure what to do about the change handler failure. tl;dr: auto-sized dimensions from `Width` / `Height` are getting persisted to the json, and the test complains about it since it's expecting nothing to change after making a change and undoing it - the change is undone correctly, but another unrelated change to `Width` / `Height` due to autosizing occurs at the same time and breaks it.

One way I guess would be to split `Size` off into `Width` / `Height`?

```diff
diff --git a/osu.Game/Skinning/SerialisableDrawableExtensions.cs b/osu.Game/Skinning/SerialisableDrawableExtensions.cs
index 0b44db9bdc..609dd9c8ab 100644
--- a/osu.Game/Skinning/SerialisableDrawableExtensions.cs
+++ b/osu.Game/Skinning/SerialisableDrawableExtensions.cs
@@ -2,6 +2,7 @@
 // See the LICENCE file in the repository root for full licence text.
 
 using osu.Framework.Bindables;
+using osu.Framework.Extensions.EnumExtensions;
 using osu.Framework.Graphics;
 using osu.Framework.Graphics.Containers;
 using osu.Game.Configuration;
@@ -19,8 +20,10 @@ public static void ApplySerialisedInfo(this Drawable component, SerialisedDrawab
             // todo: can probably make this better via deserialisation directly using a common interface.
             component.Position = drawableInfo.Position;
             component.Rotation = drawableInfo.Rotation;
-            if (drawableInfo.Size != Vector2.Zero && (component as CompositeDrawable)?.AutoSizeAxes == Axes.None)
-                component.Size = drawableInfo.Size;
+            if (drawableInfo.Width is float width && width != 0 && (component as CompositeDrawable)?.AutoSizeAxes.HasFlagFast(Axes.X) == false)
+                component.Width = width;
+            if (drawableInfo.Height is float height && height != 0 && (component as CompositeDrawable)?.AutoSizeAxes.HasFlagFast(Axes.Y) == false)
+                component.Height = height;
             component.Scale = drawableInfo.Scale;
             component.Anchor = drawableInfo.Anchor;
             component.Origin = drawableInfo.Origin;
diff --git a/osu.Game/Skinning/SerialisedDrawableInfo.cs b/osu.Game/Skinning/SerialisedDrawableInfo.cs
index 0705e91d6d..b2237acc5a 100644
--- a/osu.Game/Skinning/SerialisedDrawableInfo.cs
+++ b/osu.Game/Skinning/SerialisedDrawableInfo.cs
@@ -35,7 +35,9 @@ public sealed class SerialisedDrawableInfo
 
         public Vector2 Scale { get; set; }
 
-        public Vector2 Size { get; set; }
+        public float? Width { get; set; }
+
+        public float? Height { get; set; }
 
         public Anchor Anchor { get; set; }
 
@@ -64,7 +66,8 @@ public SerialisedDrawableInfo(Drawable component)
             Position = component.Position;
             Rotation = component.Rotation;
             Scale = component.Scale;
-            Size = component.Size;
+            Height = component.Height;
+            Width = component.Width;
             Anchor = component.Anchor;
             Origin = component.Origin;
 

```

I _think_ the above fixes it but am not 100% on this because this test seems to intermittently fail for me [earlier](https://github.com/ppy/osu/blob/74ccb0873edcbaa7af5a26973c30bc17e9ec3cf0/osu.Game.Tests/Visual/Gameplay/TestSceneSkinEditor.cs#L217), even on `master`.

bdach commented on osu.Game/Screens/Edit/Compose/Components/SelectionBox.cs:
```diff
@@ -91,6 +91,23 @@ public bool CanScaleY
+        /// <summary>
+        /// Whether vertical scaling support should be enabled.
+        /// </summary>
+        public bool CanScaleProportionally
```
I know that this name comes from me but I just realised it kinda doesn't work for beatmap editor, because there you can resize non-proportionally...

Maybe `CanScaleDiagonally` instead...?

xmldoc needs updating btw. And may be worth adding a clarification why this is separated out.

bdach commented on osu.Game/Skinning/LegacySongProgress.cs:
```diff
@@ -19,11 +19,14 @@ public partial class LegacySongProgress : SongProgress
         public override bool HandleNonPositionalInput => false;
         public override bool HandlePositionalInput => false;
 
+        public LegacySongProgress()
+        {
+            AutoSizeAxes = Axes.Both;
+        }
```
I see what the intent was here but this is a little dodgy...

Would add inline commentary about what this is doing. The hope is that as we transition to better per-case handling as discussed we can remove this and just handle more sanely as part of the component.

bdach commented on osu.Game/Overlays/SkinEditor/SkinSelectionHandler.cs:
```diff
@@ -31,8 +32,38 @@ public partial class SkinSelectionHandler : SelectionHandler<ISerialisableDrawab
+                    adjustSize = false;
+                    break;
+
+                // for edges, adjust size.
+                case Anchor.TopCentre:
+                case Anchor.CentreLeft:
+                case Anchor.CentreRight:
+                case Anchor.BottomCentre:
+                    // autosize elements can't be easily handled so just disable sizing for now.
+                    if (!allSelectedSupportManualSizing)
+                        return false;
+
+                    adjustSize = true;
+                    break;
```
This should be checking if all elements in there support sizing along the given axis, not just checking if they're completely autosized.

This being written the way it is, is probably the root cause for all of the following funky behaviours:

https://github.com/ppy/osu/assets/20418176/11026a4e-fa8d-4c14-b9dd-4d658f706051

bdach replied:
For reference, I mean something along the lines of

```diff
diff --git a/osu.Game/Overlays/SkinEditor/SkinSelectionHandler.cs b/osu.Game/Overlays/SkinEditor/SkinSelectionHandler.cs
index 77e522a925..a4d530a4d9 100644
--- a/osu.Game/Overlays/SkinEditor/SkinSelectionHandler.cs
+++ b/osu.Game/Overlays/SkinEditor/SkinSelectionHandler.cs
@@ -32,11 +32,11 @@ public partial class SkinSelectionHandler : SelectionHandler<ISerialisableDrawab
             UpdatePosition = updateDrawablePosition
         };
 
-        private bool allSelectedSupportManualSizing => SelectedItems.All(b => (b as CompositeDrawable)?.AutoSizeAxes == Axes.None);
+        private bool allSelectedSupportManualSizing(Axes axis) => SelectedItems.All(b => (b as CompositeDrawable)?.AutoSizeAxes.HasFlagFast(axis) == false);
 
         public override bool HandleScale(Vector2 scale, Anchor anchor)
         {
-            bool adjustSize;
+            Axes adjustAxis;
 
             switch (anchor)
             {
@@ -45,19 +45,25 @@ public override bool HandleScale(Vector2 scale, Anchor anchor)
                 case Anchor.TopRight:
                 case Anchor.BottomLeft:
                 case Anchor.BottomRight:
-                    adjustSize = false;
+                    adjustAxis = Axes.Both;
                     break;
 
                 // for edges, adjust size.
+                // autosize elements can't be easily handled so just disable sizing for now.
                 case Anchor.TopCentre:
+                case Anchor.BottomCentre:
+                    if (!allSelectedSupportManualSizing(Axes.Y))
+                        return false;
+
+                    adjustAxis = Axes.Y;
+                    break;
+
                 case Anchor.CentreLeft:
                 case Anchor.CentreRight:
-                case Anchor.BottomCentre:
-                    // autosize elements can't be easily handled so just disable sizing for now.
-                    if (!allSelectedSupportManualSizing)
+                    if (!allSelectedSupportManualSizing(Axes.X))
                         return false;
 
-                    adjustSize = true;
+                    adjustAxis = Axes.X;
                     break;
 
                 default:
@@ -151,10 +157,20 @@ public override bool HandleScale(Vector2 scale, Anchor anchor)
                 if (Precision.AlmostEquals(MathF.Abs(drawableItem.Rotation) % 180, 90))
                     currentScaledDelta = new Vector2(scaledDelta.Y, scaledDelta.X);
 
-                if (adjustSize)
-                    drawableItem.Size *= currentScaledDelta;
-                else
-                    drawableItem.Scale *= currentScaledDelta;
+                switch (adjustAxis)
+                {
+                    case Axes.X:
+                        drawableItem.Width *= currentScaledDelta.X;
+                        break;
+
+                    case Axes.Y:
+                        drawableItem.Height *= currentScaledDelta.Y;
+                        break;
+
+                    case Axes.Both:
+                        drawableItem.Scale *= currentScaledDelta;
+                        break;
+                }
             }
 
             return true;
@@ -203,7 +219,8 @@ protected override void OnSelectionChanged()
         {
             base.OnSelectionChanged();
 
-            SelectionBox.CanScaleX = SelectionBox.CanScaleY = allSelectedSupportManualSizing;
+            SelectionBox.CanScaleX = allSelectedSupportManualSizing(Axes.X);
+            SelectionBox.CanScaleY = allSelectedSupportManualSizing(Axes.Y);
             SelectionBox.CanScaleProportionally = true;
             SelectionBox.CanFlipX = true;
             SelectionBox.CanFlipY = true;

```

This fixes a crash, too; to see it, try resizing any health bar (the argon one will do) vertically or horizontally. It'll die due to trying to set `Size` on a half-autosized component which is not allowed.

peppy committed:
Rename diagonal scale variable and update xmldoc

peppy committed:
Add comment explaining `AutoSize` change in `LegacySongProgress`

peppy committed:
Check X/Y sizing available separately to fix weird edge cases

peppy commented:
Have applied changes 👍 .

bdach committed:
Fix test failures due to fluctuations in needlessly-serialised automatic sizings

bdach committed:
Remove argon health display bar length setting

It is no longer needed.

Intentionally not doing backwards migration to simplify things; users
can fix their skins.

bdach committed:
Merge branch 'master' into skin-size-editing

bdach committed:
Fix typo

bdach reviewed (dismissed):
I'm okay with trying this.

All of the remaining cases that are broken by this I'm explicitly ignoring for the time being. As discussed IRL the eventual goal is to have every single component handling scaling / sizing changes fully by itself, and adjusting its layout in the most correct way for itself. Can be done in separate pulls one-by-one later (probably after we introduce a more proper API to do that).

bdach committed:
Remove unused using directive

bdach committed:
Remove another reference to bar length

bdach committed:
Only store width/height of serialised drawable if it isn't automatically computed

peppy committed:
Flip comparison to allow non-composite drawables to still get resized

peppy committed:
Improve test comparison logic

Doesn't really help that much with nunit output,
but at least you can breakpoint and see the json kinda.

bdach committed:
Remove unused using directive

bdach commented on osu.Game.Tests/Visual/Gameplay/TestSceneSkinEditor.cs:
```diff
@@ -214,7 +217,11 @@ public void TestUndoEditHistory()
-            AddAssert("Nothing changed", () => defaultState.SequenceEqual(changeHandler.GetCurrentState()));
+
+            AddAssert("Nothing changed",
+                () => JsonConvert.DeserializeObject<IEnumerable<SerialisedDrawableInfo>>(Encoding.UTF8.GetString(defaultState)),
+                () => Is.EqualTo(JsonConvert.DeserializeObject<IEnumerable<SerialisedDrawableInfo>>(Encoding.UTF8.GetString(changeHandler.GetCurrentState())))
```
This can't ever work because `SerialisedDrawableInfo` doesn't implement `IEquatable<SerialisedDrawableInfo>`. So it will always be a reference comparison that will always fail.

bdach replied:
Reverted as per IRL discussion

bdach committed:
Revert test changes

They were broken because `SerialisedDrawableInfo` is a reference type
and as such equality checks will use reference equality rather than
member equality.

bdach committed:
Merge branch 'master' into skin-size-editing

bdach committed:
Use barely better assertion fail message in test

bdach committed:
Fix song progress bars not sizing vertically properly
